docs: add MVUX custom controls how-to guide (#3042)
unoplatform/uno.extensions: doc/Learn/Mvux/Tutorials/HowTo-CustomControls.md

> [!TIP]
> This guide assumes you are already familiar with [MVUX fundamentals](xref:Uno.Extensions.Mvux.Overview), [feeds](xref:Uno.Extensions.Mvux.HowToSimpleFeed), and [states](xref:Uno.Extensions.Mvux.HowToSimpleState).

## The layer separation principle

A custom control and an MVUX model serve different roles and should never be coupled directly:

| Layer | Responsibility | Knows about MVUX? |
|-------|---------------|-------------------|
| **Custom control** | Renders UI, exposes DependencyProperties, raises events | No |
| **MVUX Model** | Composes feeds/states from services, exposes commands | Yes |
| **Generated ViewModel** | Bridges the model to the data-binding engine | Auto-generated |

The custom control binds to the **generated ViewModel** through standard data binding — the same way any built-in WinUI control does. There is no special MVUX integration required inside the control itself.

## 1. Defining DependencyProperties for MVUX consumption

### Use `IEnumerable<T>` for collection properties

> When your custom control displays a collection that comes from an `IListFeed<T>` or `IListState<T>`, type your DependencyProperty as `IEnumerable<T>` — **not** `ObservableCollection<T>`:

Reviewer: > that comes from an `IListFeed<T>` or `IListState<T>`,

This is miss-understandable and unclear. The customized Control *might* be "just" binding to some Data in the Model, but what if (like in the matrix app?) the UI Layer triggers something adding up this `IEnumerable<T>` with one or more Items. Then its a TwoWay flow and this sentence does leave the Reader unsure about what to do then.

## The layer separation principle

A custom control and an MVUX model serve different roles and should never be coupled directly:

| Layer | Responsibility | Knows about MVUX? |
|-------|---------------|-------------------|
| **Custom control** | Renders UI, exposes DependencyProperties, raises events | No |
| **MVUX Model** | Composes feeds/states from services, exposes commands | Yes |
| **Generated ViewModel** | Bridges the model to the data-binding engine | Auto-generated |

The custom control binds to the **generated ViewModel** through standard data binding — the same way any built-in WinUI control does. There is no special MVUX integration required inside the control itself.

## 1. Defining DependencyProperties for MVUX consumption

### Use `IEnumerable<T>` for collection properties

When your custom control **displays** a collection owned by the Model (sourced from an `IListFeed<T>` or `IListState<T>`), type your DependencyProperty as `IEnumerable<T>` — **not** `ObservableCollection<T>`:

> [!NOTE]
> This guidance applies when the **Model owns the collection** and the control only reads it for display.

> > If your control internally manages its own collection (e.g., a game loop spawning and removing entities), the control should own a private `ObservableCollection<T>` internally and the DependencyProperty serves only as an external binding surface. See [Section 4](#4-wrapping-platform-controls-contentdialog-webview2-skcanvaselement) for imperative/rendering controls that manage their own state.

Reviewer: ```suggestion
> If your control internally manages its own collection (e.g., a game loop spawning and removing entities), the control should own a private `ObservableCollection<T>` internally and the `DependencyProperty` serves only as an external binding surface. See [Section 4](#4-wrapping-platform-controls-contentdialog-webview2-skcanvaselement) for imperative/rendering controls that manage their own state.
```

                           Difficulty="{Binding Difficulty, Mode=TwoWay}" />
</Page>
```

No special wiring is needed. The generated `GameViewModel` exposes `Loot` as a bindable collection and `Difficulty` as a bindable property. The custom control receives them through its DependencyProperties.

> [!TIP]
> MVUX auto-generates a ViewModel for each `partial class` or `record` named with a `Model` suffix. Since V5, the generated class uses the `*ViewModel` naming convention (e.g., `GameModel` -> `GameViewModel`). The older `Bindable*Model` naming from V4 is deprecated. See [Upgrading MVUX](xref:Uno.Extensions.Mvux.ReactiveMigration) for details.

### Using FeedView for loading and error states

If your control should display loading or error states, wrap it in a `FeedView`:

```xml
<uer:FeedView Source="{Binding Loot}">
    <uer:FeedView.ValueTemplate>
        <DataTemplate>
            <controls:LootDisplay Items="{Binding Data}" />
        </DataTemplate>
    </uer:FeedView.ValueTemplate>
    <uer:FeedView.ProgressTemplate>
        <DataTemplate>
            <ProgressRing IsActive="True" />
        </DataTemplate>
    </uer:FeedView.ProgressTemplate>

> </uer:FeedView>

Reviewer: @mtmattei 🤔  while I would honestly not know, how to do/define this myself, what about the idea to use the `FeedView.ErrorTemplate` for a Failed Run-Result? 👍  💡
e.g. Do we need a real Exception to happen? Where / is there a Feed/State returned Option/Message argument Property, would we be able to set a Value to? For example, in regular Event Handlers/APIException, this would/could be a `args.IsError` `args.ErrorDescription` but in MVUX State-management...?

> [!IMPORTANT]
> If your custom control **internally owns and mutates its own collections** (e.g., a game loop that spawns/removes entities on a canvas), it falls into this category — not Section 1. The control should manage its state with private `ObservableCollection<T>` fields internally and expose DependencyProperties only for configuration or external data that flows in from the Model.

### Pattern: Service-driven control registration

This pattern is used in the [Matrix app](https://github.com/mtmattei/matrix) for its `MatrixTransitionOverlay`:

**Step 1: Define a service interface**

```csharp
public interface IOverlayService
{
    Task ShowOverlayAsync(OverlayOptions options, CancellationToken ct = default);
    void RegisterOverlay(MyOverlayControl control);
}
```

**Step 2: Place the control in Shell.xaml (or a high-level page)**

```xml
<Grid>
    <Frame uen:Region.Attached="true" />
    <controls:MyOverlayControl x:Name="Overlay" />

> </Grid>

Reviewer: usually we would define `Grid.Column` or `Grid.Row` Attached Properties on the contained Controls 🤔 I can not find a information here, if this is left out on purpose in here (if yes, why?) or a issue we should fix in this sample.

// Shell.xaml.cs
private async void Shell_Loaded(object sender, RoutedEventArgs e)
{
    var host = ((App)App.Current).Host;
    var overlayService = host.Services.GetRequiredService<IOverlayService>();
    overlayService.RegisterOverlay(Overlay);
}
```

**Step 4: Call from any Model**

```csharp
public partial record GameModel(IOverlayService OverlayService)
{
    public async ValueTask ShowVictoryOverlay(CancellationToken ct)
    {
        await OverlayService.ShowOverlayAsync(new OverlayOptions(Duration: TimeSpan.FromSeconds(3)), ct);
    }
}
```

### ContentDialog with Navigation Extensions

For ContentDialog, prefer [Navigation Extensions](xref:Uno.Extensions.Navigation.HowToShowDialog) over the service pattern:

```csharp
// From a model method (INavigator injected via constructor)
await Navigator.NavigateViewAsync<ConfirmDeleteDialog>(this, qualifier: Qualifiers.Dialog);
```

Or declaratively from XAML using the `!` prefix:

```xml
<Button Content="Delete" uen:Navigation.Request="!ConfirmDelete" />
```

## 5. Accessing configuration (IOptions&lt;T&gt;) from MVUX

Configuration values should flow through the **Model** or **Service** layer, never directly into the view.

> ### Inject into the Model

Reviewer: ```suggestion
### [Inject static `IOptions<T>` into the Model](#tab/static-configuration)
```

```csharp
public partial record GameModel(
    IOptions<GameConfig> Config,
    ILootService LootService)
{
    public IState<double> Speed =>
        State.Value(this, () => Config.Value.Speed);

    public IListState<Loot> Loot =>
        ListState.Async(this, LootService.GetLootAsync);
}
```

The view binds to `Speed` like any other state:

```xml
<controls:GameBoard Speed="{Binding Speed}" />
```

> ### For runtime-changeable configuration

Reviewer: ```suggestion
### [For runtime-changeable configuration](#tab/runtime-changeable-configuration)
```

    {
        await Config.UpdateAsync(c => c with { Difficulty = level });
    }
}
```

### Prefer injecting into Services

For configuration that drives business logic (API keys, feature flags, thresholds), inject `IOptions<T>` into the service rather than the model:

```csharp
public class LootService : ILootService
{
    private readonly GameConfig _config;

    public LootService(IOptions<GameConfig> config)
    {
        _config = config.Value;
    }

    public async ValueTask<IImmutableList<Loot>> GetLootAsync(CancellationToken ct)
    {
        // Use _config.DropRate, _config.MaxItems, etc.
    }
}
```

The model stays focused on composing feeds/states, and the service owns the business rules that depend on configuration.

## Summary

| Scenario | Pattern |
|----------|---------|
| Collection DP on custom control (Model-owned data) | Type as `IEnumerable<T>`, subscribe to `INotifyCollectionChanged` |
| Collection managed internally by the control | Private `ObservableCollection<T>`, service-driven pattern (Section 4) |
| Single-value DP | Type to the entity type directly |
| Control events -> Model | Code-behind handler, `CommandExtensions`, or command DP |

> | Rendering surfaces (WebView2, SKCanvas) | Service-driven registration pattern |

Reviewer: ```suggestion
| Rendering surfaces ( [`WebView2`](https://platform.uno/docs/articles/controls/WebView.html), [`SKCanvasElement`](https://platform.uno/docs/articles/controls/SKCanvasElement.html) ) | Service-driven registration pattern |
```

### The XAML

```xml

> <Page x:Class="MyApp.Presentation.GamePage"

Reviewer: We should expect AI/MCP driven agents, reading this guide to assume this and other code-samples as complete working sample - which would potentially not be the case like this. Usually our page files in Uno already come with >4-5 `xmlns`, *including one with `local` (codebehind?), so this here here looks like in-complete to me.

      xmlns:controls="using:MyApp.Controls">

    <controls:LootDisplay Items="{Binding Loot}"
                           Difficulty="{Binding Difficulty, Mode=TwoWay}" />
</Page>
```

No special wiring is needed. The generated `GameViewModel` exposes `Loot` as a bindable collection and `Difficulty` as a bindable property. The custom control receives them through its DependencyProperties.

> [!TIP]

> > MVUX auto-generates a ViewModel for each `partial class` or `record` named with a `Model` suffix. Since V5, the generated class uses the `*ViewModel` naming convention (e.g., `GameModel` -> `GameViewModel`). The older `Bindable*Model` naming from V4 is deprecated. See [Upgrading MVUX](xref:Uno.Extensions.Mvux.ReactiveMigration) for details.

Reviewer: ```suggestion
MVUX auto-generates a ViewModel for each `partial class` or `partial record` named with a `Model` suffix. Binding from UI to this generated ViewModel, uses same names for the bound Properties like we are defining in the MVUX Model, as name of the `IListState<Loot> Loot` / `IState<DifficultyLevel> Difficulty`.
```
🤔 @mtmattei for this part: `Since V5, the generated class uses the `*ViewModel` naming convention (e.g., `GameModel` -> `GameViewModel`). The older `Bindable*Model` naming from V4 is deprecated. See [Upgrading MVUX](xref:Uno.Extensions.Mvux.ReactiveMigration) for details.` this would be quite unconventional in Uno Docs, to tell about this in Guides like this here. My Comment was more meant to stay compliant to the latest version way-to-do-this. *If* we still want to link the version differences, this should potentially only be located on the very bottom in a kind of "Future Readings" 🤷 

Trying around with copilot + my App, it came up with the remark, we should/can not define the Property equally named like the Type `Loot` for example. Can you check and validate this does not cause problems or needs special implementations/workaround?

One thing to the `Difficulty`: When would we define this as `IFeed<DifficultyLevel>` instead and call a Task/ValueTask on the Service or maybe we should define a `PauseDialogService`?
is there a difference in usage of one or the other (Task/ValueTask), with/in MVUX? when should we realistically which of them, beside from Heap/Stack? The Uno MVUX Docs would not show me a clear indicator and both are possible from my POV

> Platform controls that are **rendering surfaces** (WebView2, SKCanvasElement) or **modal UI** (ContentDialog) require a different pattern than data-bound controls. They do not fit the standard DP-binding model because they are driven by imperative APIs.

Reviewer: ```suggestion
Platform controls that are **rendering surfaces** ( [`WebView2`](https://platform.uno/docs/articles/controls/WebView.html), [`SKCanvasElement`](https://platform.uno/docs/articles/controls/SKCanvasElement.html) ) or **modal UI** ( [`ContentDialog`](https://platform.uno/docs/articles/features/dialogs.html?tabs=mvux%2Cxaml) ) require a different pattern than data-bound controls. They do not fit the standard DP-binding model, because they are driven by imperative APIs.
```
the Reader of this here might not be in-depth with words like `imperative` (end of the last sentence) but checking Google, this might be a good thing he needs/should be made known to.

Quote from [`react-spring`](https://share.google/CEmay3cjDbMGjjN0l)  cursive formated part I got highlighted, but the second sentence might also be relevant to know.

> The imperative API *enables you to update your animations without requiring a react-render to occur*. This is useful for animations that are not tied to a component's lifecycle, such as animations that are triggered by user input.

lines 222-225 are a bit high level/difficulter to grasp and should be made a bit easyer to understand/simpler worded?

> [!IMPORTANT]

> > If your custom control **internally owns and mutates its own collections** (e.g., a game loop that spawns/removes entities on a canvas), it falls into this category — not Section 1. The control should manage its state with private `ObservableCollection<T>` fields internally and expose DependencyProperties only for configuration or external data that flows in from the Model.

Reviewer: could you maybe add a sample, how this DP + Callback on change would then look like, with `ObservableCollection<T>` and how to connect & cast the (new?) Collection properly ?
Possible Problem: A Developer did never learn Mvvm itself, then started directly with MVUX entusiastic about, not having to learn the ObservableCollection / INotifyPropertyChanged Way-to-do-it at all, and will because of this not be directly able to validate/see miss-implementations / bad practice implementations done by the Agent fed by incomplete instructions 🤷 
If we know a offical MS Docs resource providing this as sample, then please add a link to this! Please carefully check *if we* would want to recommend Mvux + Mvvm UnoFeature usage side by side for customized controls like this, before causing the Reader of this guide to stumble above the next Obstacle to handle. I learned from a comment from @jeromelaban  last year, that we normally should not combine both Features as it can cause problems with the Generators

| Scenario | Pattern |
|----------|---------|
| Collection DP on custom control (Model-owned data) | Type as `IEnumerable<T>`, subscribe to `INotifyCollectionChanged` |
| Collection managed internally by the control | Private `ObservableCollection<T>`, service-driven pattern (Section 4) |
| Single-value DP | Type to the entity type directly |

> | Control events -> Model | Code-behind handler, `CommandExtensions`, or command DP |

Reviewer: @mtmattei this looks wrong 🤔 Could you validate this?
Especially this here: `Control events -> Model | Code-behind handler,` would allow/motivate the, already many times misslead, Agent to assume our before shown codebehind sample:

https://github.com/unoplatform/uno.extensions/pull/3042/changes/a37a5c97a45ff41c4a6158d447cf49a4f9bef62e..da740836cb425a06e533ddfdacc30072446f895b#diff-20162b1b6e2ce8875ceed8644fcc8d70200f52b72022dcbf2251f8ee3a3032a6R180

to be *one-way* and a direct access to `GameModel` in codebehind would be valid, which *is not*
